spec/client/ics-028-wasm-client/README.md

```diff
@@ -0,0 +1,451 @@
+
+### Headers
+
+Contents of wasm client headers depend upon `WASM Client Code`.
+
+```typescript
+interface Header {
+  data: []byte
+}
+```
+
+### Misbehaviour
+
+The `Misbehaviour` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
+WASM client `Misbehaviour` consists of two headers at the same height both of which the light client would have considered valid.
+
+```typescript
+interface Misbehaviour {
+  fromHeight: Height
+  h1: Header
+  h2: Header
+}
+```
+
+### Client initialisation
+
+WASM client initialisation requires a (subjectively chosen) latest consensus state interpretable by the target WASM Client Code. `wasmCodeId` field is unique identifier for `WASM Client Code`, and `initializationData` refers to opaque data required for initialization of the particular client managed by that WASM Client Code.
+
+```typescript
+function initialise(
+    wasmCodeId: String,
+    initializationData: []byte,
+    consensusState: []byte,
+  ): ClientState {
+    codeHandle = getWASMCode(wasmCodeId)
+    assert(codeHandle.isInitializationDataValid(initializationData, consensusState))
+    set("clients/{identifier}/consensusStates/{height}", consensusState)
+    return codeHandle.initialise(initializationData, consensusState)
+}
+```
+
+The `latestClientHeight` function returns the latest stored height, which is updated every time a new (more recent) header is validated.
+
+```typescript
+function latestClientHeight(clientState: ClientState): Height {
+  codeHandle = clientState.codeHandle();
+  codeHandle.latestClientHeight(clientState)
+}
+```
+
+### Validity predicate
+
+WASM client validity checking uses underlying WASM Client code. If the provided header is valid, the client state is updated & the newly verified commitment written to the store.
+
+```typescript
+function checkValidityAndUpdateState(
+  clientState: ClientState,
+  epoch: uint64,
+  header: Header) {
+    codeHandle = clientState.codeHandle()
+    isValid, consensusState = codeHandle.validateHeaderAndCreateConsensusState(clientState, header, epoch)
+    set("clients/{identifier}/consensusStates/{header.height}", consensusState)
+    // save the client
+    set("clients/{identifier}", clientState)
+}
+```
+
+### Misbehaviour predicate
+
+WASM client misbehaviour checking determines whether or not two conflicting headers at the same height would have convinced the light client.
+
+```typescript
+function checkMisbehaviourAndUpdateState(
+  clientState: ClientState,
+  misbehaviour: Misbehaviour) {
+    codeHandle = clientState.codeHandle()
+    consensusState = get("clients/{identifier}/consensusStates/{misbehaviour.fromHeight}")
+    assert(codeHandle.handleMisbehaviour(clientState, consensusState, misbehaviour))
+    // save the client
+    set("clients/{identifier}", clientState)
```

Comment:
ditto, can we give the WASM code limited state read/write access instead, it would be more flexible